### ISSUE TITLE
Futhark does not work on Windows.

### ISSUE THREAD

**Issue by mrakgr** [HIPERFIT/futhark#158]
I tried it. The `futhark-c` cannot find a header file and `futhark-opencl`'s parser goes into an infinite loop. To be honest I am not sure if running a Linux application is supposed to give an error message or if the Haskell compiler did try to produce a Windows binary - I am guessing it is the later - but it does not work at the moment.

It might be good to note this on the main page for the time being.

**Comment by Athas** [comment]
Thank you for the report.  I am not surprised that there are problems,
given that nobody here has tried to run it on a Windows machine, but I
am surprised at how they manifest themselves.

The generated C code includes headers that I think are POSIX-specific
(errno,h, sys/time.h, getopt.h, etc).  That can be fixed, or maybe
replaced with portable alternatives.

I am more surprised that `futhark-opencl` goes into an infinite loop
while parsing.  The Futhark compiler itself should not have any platform
dependent code, and the parser is exactly the same as in `futhark-c`.
Could you check that it's not actually the final linking stage (which
invokves gcc with possibly platform-dependent flags) that fails?


-- 
\  Troels
/\ Henriksen

**Comment by Athas** [comment]
Actually, could you provide the error message you get?  I'm not entirely sure which header file is missing.

**Comment by mrakgr** [comment]
Before I do that I have another question.
```
fact.c:18:21: fatal error: CL/cl.h: No such file or directory
   #include <CL/cl.h>
```
I am guessing that I am missing OpenCL which is why the above is failing on Ubuntu 14.04. Was OpenCL supposed to come with Cuda? When I installed 7.5 to run Tensorflow yesterday it actually crashed the OS and I had to scramble around to manually remove the offending packages. OpenCL might have gotten lost in the confusion. Or does it maybe come with a separate package? I cannot find anything with apt-get though.

I'll fetch the Windows error messages momentarily...

**Comment by mrakgr** [comment]
```
gcc failed with code 1:
fact.c:10:17: fatal error: err.h: No such file or directory
 #include <err.h>
                 ^
compilation terminated.
```
The above is for both `futhark-c` and `futhark-opencl`. I remembered incorrectly.

`futharki` goes into an infinite loop of...
```
Left Parse error: End of file
Left Parse error: End of file
Left Parse error: End of file
...
```
...error messages.

**Comment by Athas** [comment]
I have seen CUDA being installed in /usr/local before, in which case the include files are not in the default include path.  You can fix it by prepending `/usr/local/cuda/include` to your CPATH:

```
export CPATH=/usr/local/cuda/include:$CPATH
```

(I don't know why this varies between systems.)

With respect to `futharki`, yeah, the parser broke at some point.  Don't worry though, you're not missing much: the interactive form of `futharki` just spits out the AST of whatever expression you entered.  It was never actually finished as a useful REPL.  I'll make an issue for it.

**Comment by Athas** [comment]
Oh, and I'll try to think of a way to fix that Windows compilation issue.

**Comment by mrakgr** [comment]
I just realized that `futharki` is broken on Linux as well.

**Comment by Athas** [comment]
I just pushed a fix that removes the use of the nonstandard err.h header.  Could you check that it works on Windows?

**Comment by mrakgr** [comment]
Ok. Also your suggestion to add the Cuda directory to CPATH worked. Right now I am trying to wrap my head around some of the examples. I'll check whether it works on Windows.

Right now, Linux seems to be unstable so if Futhark could be made to work on Windows, I'd like it very much.

**Comment by Athas** [comment]
I also just pushed a fix for the futharki issue, but as mentioned, it's still not a usable tool.

**Comment by mrakgr** [comment]
Ok, `futhark-c` work on Windows now. `futhark-opencl` spits out this error message:
```
gcc failed with code 1:
fact.c: In function 'opencl_succeed':
fact.c:133:5: warning: implicit declaration of function 'errx' [-Wimplicit-function-declaration]
     errx(-1, "%s:%d: OpenCL call\n  %s\nfailed with error code %d (%s)\n",
     ^
fact.c: At top level:
fact.c:582:1: error: unknown type name 'suseconds_t'
 static suseconds_t reduce_kernel_97total_runtime = 0;
 ^
fact.c:585:1: error: unknown type name 'suseconds_t'
 static suseconds_t reduce_kernel_110total_runtime = 0;
 ^
fact.c: In function 'futhark_main':
fact.c:1533:13: error: unknown type name 'suseconds_t'
             suseconds_t time_diff_141 = time_end_140.tv_sec * 1000000 +
             ^
fact.c:1581:13: error: unknown type name 'suseconds_t'
             suseconds_t time_diff_146 = time_end_145.tv_sec * 1000000 +
             ^
fact.c: In function 'main':
fact.c:1974:5: error: unknown type name 'suseconds_t'
     suseconds_t total_runtime = 0;
     ^
fact.c:1975:5: error: unknown type name 'suseconds_t'
     suseconds_t total_runs = 0;
     ^
```
At first it could not find the `cl.h` file, but after I added it to path the above is happening. I am using the GCC 4.8.0 compiler.

**Comment by Athas** [comment]
I got rid of the final calls to `errx()` and replaced `suseconds_t` with `int`.  It might work now.

**Comment by mrakgr** [comment]
Hmmm, I can't synchronize my fork with this one. Give me a bit...

**Comment by mrakgr** [comment]
```
gcc failed with code 1:
c:/x86_64-w64-mingw32-gcc-4.8.0-win64_rubenvb/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/4.8.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lOpenCL
collect2.exe: error: ld returned 1 exit status
```

This is what I get now.

**Comment by mrakgr** [comment]
I tried adding `C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v7.5\lib\x64` where the `OpenCL.lib` is located to CPATH, but no such luck. Maybe I need to add the directory to some other variable?

**Comment by mrakgr** [comment]
I though that adding the directory to `LD_LIBRARY_PATH` might work, but no such luck. Is it trying to load the `.so` version of the library maybe?

**Comment by Athas** [comment]
I'm not sure what the Windows version of GCC looks for.  I'd guess a DLL.

**Comment by Athas** [comment]
There are some complicated instructions for how to link to OpenCL here: https://stackoverflow.com/questions/15185955/compile-opencl-on-mingw-nvidia-sdk

I have a hard time believing it's really supposed to be this complicated, though.

**Comment by Athas** [comment]
According to [this guide](https://saintdako.github.io/blog/2015/07/09/windowsOpenCL/), what you are doing should essentially work.  Are you sure that you modified CPATH correctly?  (I have no idea where that is done on Windows.)  Maybe there is an issue where the backslashes are interpreted by the shell, thus mangling the file name?

**Comment by mrakgr** [comment]
I'll give that guide a try. As to where I put the `CPATH`, it was in the environment variables. Windows has something similar to Linux in the system options. I'll double check it.

**Comment by mrakgr** [comment]
I followed the instructions in that guide and the program compiles. When I try to run it though, it crashes without any error message. I'll try going through it in a debugger, though I am not sure what to do at this point. Maybe I'll try compiling the example in Visual C++.

**Comment by Athas** [comment]
Maybe you have to log in and out for the environment variables to take effect.  It used to be that way in Windows XP I think - how much can things change in a mere fifteen years? (Also, double-check the quoting rules.)

I think to run the program, you need to make sure that the OpenCL DLL is available - for example by putting it in the same directory as the program.  There is probably also some sort of include path for DLLs, but this is beyond my knowledge of Windows.

Does `futhark-c` at least work properly?

**Comment by mrakgr** [comment]
Yes, `futhark-c` does. 

Yeah, the debugger does give me `proc_not_found`, so maybe your idea of putting the OpenCL dll into the directory might have merit. In Windows 10, you do not need to log in and out, just apply the changes and restart the command prompt.

I'll look into locating the `opencl.dll` now.

**Comment by mrakgr** [comment]
As it turns out there are a few in my system, including the Windows directory and I have no idea why one the program is trying to use. It could be using the 32 bit version incorrectly for all I know. The OpenCL version used by Nvidia is apparently also out of date.

Erm, let me try to get anything in OpenCL working whether in VS or GCC and I'll trace my steps to here from there.

**Comment by Athas** [comment]
Right, NVIDIA only supports 1.2 or thereabouts.  That should not be a problem for the simple example program, and it is certainly not a problem for Futhark (our department only really has NVIDIA hardware, unfortunately).

It may be that the program runs succesfully (well, it always exit(1)s for some reason), but that it simply does not find any OpenCL devices, and thus produces no output.  OpenCL is really a wrapper library around things called "ICD"s, which are the real device-specific drivers.  On Linux, these ICDs are stored in /etc/OpenCL/vendors, but I have no idea where they may be on Windows.

**Comment by mrakgr** [comment]
I can confirm that the example in the guide compiles and runs just fine in Visual C++.

I tried putting in the `fact.c` produced by `futhark-opencl` into VC, but they aren't compatible. VC flags the implicit conversions from `void *` to `char *` as errors. I'll try looking around and asking on Stack Overflow next for lack of better option.

**Comment by Athas** [comment]
mrakgr <notifications@github.com> writes:

> I can confirm that the example in the guide compiles and runs just fine in Visual C++.
>
> I tried putting in the `fact.c` produced by `futhark-opencl` into VC,
> but they aren't compatible. VC flags the implicit conversions from
> `void *` to `char *` as errors. I'll try looking around and asking on
> Stack Overflow next for lack of better option.

Alright.  I'll see if I can find those casts and make them explicit.  I
guess they are because Visual C++ uses a C++ compiler, and C++ has
stricter type rules?

-- 
\  Troels
/\ Henriksen

**Comment by mrakgr** [comment]
I do not have much experience with GCC, but that is what I have read.

Also `#include <sys/time.h>` and `#include <getopt.h>` cannot be found, so the methods that use the stuff from that header are also flagged as type errors.

**Comment by Athas** [comment]
Well, I have pushed a patch that makes Futhark generate valid C++.  That may come in handy someday for other reasons.  However, if Visual C++ does not provide `sys/time.h` and `getopt.h`, then I am skeptical about getting it to work.  Replacing those would be much more work than the fixes I've committed so far.

I can only suggest finding out why the MinGW toolchain fails to build OpenCL programs.  As far as I can see, Futhark now does what it's supposed to do (although I'm not opposed to tacking on more compiler flags on Windows, if necessary).

**Comment by mrakgr** [comment]
I'll try using the `reimp` tool as soon as I figure out how to get it. That should work. It seems that MingW needs to compile directly against the dll or something like that.

**Comment by mrakgr** [comment]
Ok, I've tried everything from the first SO post. I've installed `mingw-utils`, found the `reimp` tool, tried to use it on the `OpenCL.lib`, but it failed unfortunately with an invalid or corrupt library error. After that I went for the alternative and used the `dlltool` on the `.def` file that I found in one of the posts. I finally figured out how to link to `libOpenCL.a`, but it seems that the library has changed in some way as I get the following error:
```
C:\futhark\examples>g++ -std=c++0x -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v7.0\include" main.cpp -L"C:\futhark\examples" -lOpenCL -o main.exe
C:\Users\Marko\AppData\Local\Temp\ccA4Tua5.o:main.cpp:(.text+0x10f): undefined reference to `clGetPlatformInfo'
C:\Users\Marko\AppData\Local\Temp\ccA4Tua5.o:main.cpp:(.text+0x165): undefined reference to `clGetPlatformInfo'
C:\Users\Marko\AppData\Local\Temp\ccA4Tua5.o:main.cpp:(.text+0x1b0): undefined reference to `clGetDeviceInfo'
c:/x86_64-w64-mingw32-gcc-4.8.0-win64_rubenvb/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/4.8.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\Marko\AppData\Local\Temp\ccA4Tua5.o: bad reloc address 0x0 in section `.pdata$_ZNSt11char_traitsIcE6assignERcRKc'
collect2.exe: error: ld returned 1 exit status
```
So close and yet so far. That is cross platform development for you. I think I did all I could with my current knowledge. I think that I'll switch to asking around unless you have any good ideas.

**Comment by Athas** [comment]
Wow, that is pretty crazy.  That 'bad reloc address 0x0' sounds really
scary.  The OpenCL functions it complains about are very foundational,
so I have a hard time believing that they have not been included.

Thank you for not giving up yet!  There is in principle no reason why
Futhark should not work on Windows, and I'd very much like for it to be
the case.  I'll look around for more information on MinGW/OpenCL, but
I'm a bit hampered by no having a Windows system available.  I wonder if
I could Wine it...

-- 
\  Troels
/\ Henriksen

**Comment by mrakgr** [comment]
I did a post on MinW-Users mailing list, but just in case that falls through, would it be possible to rewrite the `sys\time.h` related stuff to use platform independent C++ functions?

Actually, there are two options: here is the stuff from the [C standard library](http://www.cplusplus.com/reference/ctime/) and here are the [newer C++ template based clocks and timers](http://www.cplusplus.com/reference/chrono/).

**Comment by Athas** [comment]
That approach is unfortunately not feasible, since Futhark generates C,
not C++.  But maybe there is some standardised C stuff for timing now?
We can always embed a getopt.h-implementation in the generated code
instead of importing it - it's fairly simple code.

I have a fairly high tolerance for ugliness in generated code, but I
have a low tolerance for significant bodies of code that are only active
on one platform - especially when that platform is not part of our
automatic test suite.  If Travis supported Windows, this would be
different.

-- 
\  Troels
/\ Henriksen

**Comment by mrakgr** [comment]
The stuff in `<ctime>`/`time.h` should be standardized unless I am badly mistaken. Regarding `getopt.h` I already found a Windows implementation on the net. It did seem to be simple enough.

**Comment by Athas** [comment]
It looks like time.h is a useful substitute for sys/time.h.  I'll give
it a try.

-- 
\  Troels
/\ Henriksen

**Comment by Athas** [comment]
Hm, maybe not: time() only supports second-level granularity, and
clock() counts processor time, where we need wall time.

-- 
\  Troels
/\ Henriksen

**Comment by mrakgr** [comment]
How about the first answer to this [question](http://stackoverflow.com/questions/17432502/how-can-i-measure-cpu-time-and-wall-clock-time-on-both-linux-windows)?

**Comment by Athas** [comment]
Yeah, I guess something like that is necessary.  I just have a really
bad feeling about #ifdef'd code that is never run on our test systems.

-- 
\  Troels
/\ Henriksen

**Comment by Athas** [comment]
I guess it's small enough that it may be robust... if I fix this and also include the moral equivalent of `stdarg.h`, do you think that's enough to support Windows?

**Comment by mrakgr** [comment]
I can't say, I do most of my programming in F# and the stuff there is mostly plug and play. I would really like to see Futhark succeeding because at the current time Cuda programming is such a major bottleneck in [my own work](https://github.com/mrakgr/Spiral-Q). I think that for every line of F# that I write it would take an equivalent 30 lines of Cuda C++. A compiler library like [Alea](http://quantalea.com/) is definitely better than sending text string to NVRTC like I am doing now as it is integrated with the type system, but in the end it is still raw C with a polish.

If we can get this working on Windows, I promise that I'll do my part in testing it out and contributing to it.

**Comment by Athas** [comment]
Well, the best solution to that problem is probably to implement an F# backend for Futhark. :-) That's actually on my medium-term list, since we use F# at a bunch of courses at my department.

But I think we need to take a step back here wrt. to the C backend on Windows.  There are really two problems that must be solved:

1. Making `futhark-opencl` able to compile standalone Futhark programs on Windows.

2. Making the generated code compilable in Visual C++.

The former is more important in the short term.  To fix it, we don't need to fix the use of `sys/time.h` or `getopt.h`, but rather figure out what's necessary to make MinGW GCC able to link with OpenCL.  Of course, one solution (that I must admit I'm not a big fan of) is to change `futhark-opencl` to invoke the Visual C++ compiler instead of `gcc`.  I would find that an acceptable solution, although I would prefer to also support GCC.

**Comment by mrakgr** [comment]
Personally, although I would like to see Futhark on Windows, it is not that I prefer GCC over VC++ or vice versa. VC++ just happens to be working at and I have no idea how to get it running with GCC.

...I am kind of curious about what courses you are doing with F#? Is it compilers? I did two of those although they were relatively simple compared to Futhark. The first one was when for practice I did the entirety of the [Compilers course](https://www.coursera.org/course/compilers) last year (I am still waiting for that one to resume so I can submit the solutions and get some virtual points) and the second when I ported half of the Java based [GVGAI library to Fsharp](https://github.com/mrakgr/GVGAI-Fsharp).

Although you joked about it, I really will create a backend for Futhark if I like it enough. I've done crazier things in programming so far.

**Comment by Athas** [comment]
We recently moved from SML to F# for introductionary programming, and
will move the compilers course likewise.

I also have no real strong opinion on GCC versus VC++ on Windows, except
that GCC is easier to obtain.  To be honest, I hope that you GCC
troubles are an isolated configuration problem.  I'll see about setting
up Windows in a virtual machine to see whether I can figure it out.

-- 
\  Troels
/\ Henriksen

**Comment by Athas** [comment]
You know, if you can come up with a command line for compiling OpenCL programs on Windows using the VC++ compiler, I'll just make `futhark-opencl` use that.

**Comment by nqpz** [comment]
Hello, I'm just joining the conversation!

I have access to a fairly basic Windows 8.1 (no Visual Studio or IDE or anything much except a web browser installed), so I thought I would try to install Futhark on that.  I have succeeded using GCC, but OpenCL is still annoying.

Once I got `ghc` and `futhark` up and running through Stack, I installed MinGW and added its `C:\MinGW\bin` directory to the system PATH environment variable (this might not be necessary, but it enabled me to use MinGW's executables outside of MinGW's MSYS shell).  After these steps, I could run `futhark-c` with no problems, and also run the generated programs.

Then I went on to see if I could get `futhark-opencl` to work.

Initially GCC complained about missing headers, so I fetched `cl.h` and `cl_platform.h` from Khronos' website.  The computer uses a nVidia GPU with support for OpenCL 1.2, so I picked the headers from that version.  I created a new directory `C:\MinGW\include\CL` and put the two header files in there.

GCC then stopped complaining about missing headers.  Next up it yelled that there was no OpenCL library to be found.  I looked a bit around in `C:\Program Files` and found a directory named `NVIDIA Corporation`.  Inside that was a directory named `OpenCL`, which had a file named `OpenCL.dll`.  This sounded legit, so I copied it into `C:\MinGW\lib`.  (I'm not a Windows expert, but copying DLL files around seems about right.)

After all of these steps I was able to run `futhark-opencl` on several `.fut` files and run the generated executables and get the correct results.  The generated executables were faster than those generated with `futhark-c`, so it seems like they were actually running on the GPU.

In conclusion, it seems like I got it working with GCC, but only by manually downloading two header files into a manually created directory, and copying a DLL file.  This obviously isn't portable, but I don't know what could be done differently if we want to be independent of things like VC++.

(I also tried `futhark-pyopencl`, but apparently PyOpenCL requires Microsoft Visual C++, which apparently is not part of the Windows 8.1 I was using.)

**Comment by mrakgr** [comment]
@nqpz Thank you for trying. I tried to replicate what you did. I copied the relevant header files into `C:\x86_64-w64-mingw32-gcc-4.8.0-win64_rubenvb\mingw64\include` and the dlls into `C:\x86_64-w64-mingw32-gcc-4.8.0-win64_rubenvb\mingw64\lib`. I also added those two directories to CPATH and LD_LIBRARY_PATH respectively. I actually had no idea where Nvidia's OpenCL dlls were located or that they needed to be linked against directly. I only looked in the GPU computing toolkit directory as that is where all the libraries and the include were. I also missed it during the file search.

```
C:\futhark\examples>g++ -std=c++0x main.cpp  -o  main.exe -lOpenCL
c:/x86_64-w64-mingw32-gcc-4.8.0-win64_rubenvb/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/4.8.0/../../../../x86_64-w64-mingw32/bin/ld.exe: skipping incompatible c:/x86_64-w64-mingw32-gcc-4.8.0-win64_rubenvb/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/4.8.0/../../../../lib/OpenCL.dll when searching for -lOpenCL
...(This error message repeats 10 times before aborting)
```

There still seem to be some incompatibilities somewhere. Maybe it would be worth it for me to actually install the freshest GCC. Also, are using the 64-bit MingW or 32-bit. I'd be surprised if 32-bit was working.

One thing that is suspicious to me here is why is it trying to link against `OpenCL.dll` when `OpenCL64.dll` seems more sensible. Let me try that one....

```
Using platform: NVIDIA CUDA
Device 0: GeForce GTX 970
```

...Ah, it worked! I just changed `-lOpenCL` to `-lOpenCL64` and now it is fine.

Also, it seems `futhark-opencl` is trying to link with the 32-bit `OpenCL.dll` on my system and it is giving me exactly the same error message as in the first code block - that it is skipping the incompatible dll.

With this, now it is entirely clear to me how to link against OpenCL and that particular issue seems solved.

The only thing that has left me wondering is why are the [instructions in the guide](http://saintdako.github.io/blog/2015/07/09/windowsOpenCL/) working for the author when he is linking against the 32-bit library using a 64-bit version of the compiler. Maybe the newer GCC 5.1.0 is smarter about linking that 4.8.0?

**Comment by Athas** [comment]
Yes, GCC 4.8.0 does seem a little old.  Is it customary to suffix library name with 64 on Windows?  I do not think I have seen that before.

In any case, I am going to close this issue.  Futhark is able to generate code that runs on Windows, assuming sufficient configuration (or hacking...) of the Windows environment.  It does not seem like we can change Futhark to improve matters across all Windows installations - changing lOpenCL to -lOpenCL64 would probably break on @nqpz's install.  The best course of action is probably to add some documentation about this issue - either to the README, or to the docs themselves.

If VC++ support is still useful, you can open another issue.

**Comment by mrakgr** [comment]
Out of interest, I am doing some research into how to get this running on `cl` (VC++ compiler) from the command line and I'll probably open an issue with the instructions just for completeness. I'll also try installing the 32-bit MinW and seeing if that patches up the issue with the default dll. I think the later should work fine though.

**Comment by mrakgr** [comment]
I did it. Let me deposit it all together here. No need to open a separate issue.

First off come the batch script to easily invoke `cl`:
```
@echo off
%comspec% /c ""c:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat" amd64 && cl.exe %*"
```

Based on this [SO post](http://stackoverflow.com/questions/7865432/command-line-compile-using-cl-exe), I made a cl.bat file and put the above in. This allows me to easily invoke the cl.

Then I invoke the `cl.bat` using something like this from the command line:

`cl main.cpp OpenCL.lib /EHsc /I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v7.5\include" /link /LIBPATH:"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v7.5\lib\x64"`

`main.cpp` is the name of the file being compiled, `OpenCL.lib` is the 64-bit library, `/link` passes the `/LIBPATH` option to the linker. The include and the library path options can be defined in the `INCLUDE` and `LIBPATH` environment variables as well, similar to MingW. 

`/Ehsc` is just a option for exception handling without which the compiler prints out a warning.

**Comment by mrakgr** [comment]
One thing the `cl` compiler warned me the first time around when is that it needs to compile in 64-bit mode to be compatible with `OpenCL.lib`. For Windows, it might be a good idea to pick the 64-bit OpenCL dll just to make sure that any unexpected bugs do not crop up.

Is it compiling 32-bit or 64-bit on Linux?

**Comment by Athas** [comment]
It is compiling in whatever `gcc` is.  32-bit on 32-bit systems, and 64-bit on 64-bit systems.  My intention has been to avoid getting the Futhark compiler too involved with exactly how the C code is being compiled, but maybe that's not an option on Windows.  I'm considering adding a `--source` flag that skips the final compilation stage, and leaves it up to the user to figure out how to get the generated C program to compile.

**Comment by mrakgr** [comment]
Could you add a flag to call `-lOpenCL64` instead of '-lOpenCL' as well? That would solve the issue on my system. Right now I'll give it a shot using 32-bit GCC, but even if that works on the simple example, it might fail on larger ones.

**Comment by Athas** [comment]
I wouldn't like that very much, as it is a global solution to a local problem.  Would it help if you just renamed OpenCL64.dll to OpenCL.dll and made sure that your renamed version is the first one that is found?

**Comment by Athas** [comment]
Hm, do you think [https://stackoverflow.com/questions/23275675/32-bit-gpu-and-64-bit-cpu-opencl](this) is related?

**Comment by mrakgr** [comment]
Well, it probably related. 

Also, I just finished trying it out with the 32-bit MingW. Now that I have it set up properly with the two dlls in the lib directory and the header files in include and the environment variables pointing to them, `futhark-opencl` works on the factorial example.

I should emphasize that this is not the correct behavior necessarily even though it works for @nqpz and myself.

The library that comes with Cuda is clearly 64-bit and incorrectly linking to the 32-bit dll is just asking for trouble. In fact rather than doing a flag, trying to link to the standard `OpenCL.dll` should be considered an error on Windows. `OpenCL64.dll` should be the default library.

Renaming the `OpenCL64.dll` to `OpenCL.dll` would be a solution, but if I was doing compiler, I'd rather not make the user go through such a convoluted step. The compiler should be smart enough to know which OS it is on and do it automatically.

On Linux, I just checked and found that there is only a single `OpenCL.a` file so there is no confusion there. I have no doubt that it is the 64-bit version.

**Comment by Athas** [comment]
That's because on Linux, the default linker path includes /usr/lib,
which contains only libraries of the native word size.  32-bit libraries
are all in /usr/lib32 (which is just called /usr/lib on a 32-bit
system).

I really am uncertain about how best to handle this.  For example, we
link succesfully with the math library on Windows using just -lm - no
-lm64 is necessary.  And what about Intel and AMD OpenCL SDKs - do these
also suffix 64 to the DLL name?  And isn't there also an OpenCL.dll
bundled with Windows (which is presumably 64-bit on 64-bit Windows)?
Why isn't that found instead of the one in the NVIDIA directory?

Since you are probably the only Futhark-on-Windows user, I'd be happy to
conditionally change -lOpenCL to -lOpenCL64 for now.  But I can't
promise it won't change if we at some point figure out exactly what is
going on. :-)

-- 
\  Troels
/\ Henriksen

**Comment by nqpz** [comment]
I should note that there was indeed also a `OpenCL64.dll` file on my Windows 8.1.  I had downloaded the 32-bit version of MinGW for no particular reason, so I didn't use it, but in general it's probably true that you're going to use the 64-postfixed variant.  But I'm still not sure if it's only nVidia who produces `OpenCL64.dll` files, or if other GPU manufacturers use different naming schemes?  It seems a bit like the wild west on Windows when you don't use things like VC++.

**Comment by mrakgr** [comment]
Is that math library a `.dll` or an `.a` file? There is probably a difference between the two. I have no idea about the AMD and Intel SDKs. And I can confirm that there are OpenCL dlls in the Windows directory, but those are probably for the CPU.

I have to remind you that the way Nvidia intended OpenCL to be used on Windows is with the `.lib` file, not directly against the `.dll` like we are doing now. In fact, until @nqpz pointed it out, it did not even occur to me that static linking directly using dlls was even possible. So at this point linkage behavior is already anomalous.

On Windows, `.dll` are using for dynamic linking. Using dlls the program is supposed to invoke the functions inside at runtime.

**Comment by Athas** [comment]
It's the same on Linux - the Futhark compiler attempts to dynamically
link with OpenCL (that's what the -lOpenCL part does).  Is a Windows
.lib for dynamic or static linking?

Also, I've gotten a bit confused by all the details - would changing
-lOpenCL to -lOpenCL64 (on Windows) make your life appreciably easier?
If so, I'll do it.

-- 
\  Troels
/\ Henriksen

**Comment by mrakgr** [comment]
A windows `.lib` is definitely for static linking while a `.dll` is for dynamic. I am sure of this much. So Futhark is doing dynamic linking...

Well regardless, please change it to `lOpenCL64` on Windows. Thank you. 

This is to restrict it to 64-bit compilation. Cuda is 64-bit entirely by now, and most modern computers are 64-bit as well. Unless you want to explicitly support 32-bit OpenCL then this is the way to go. I do not think adding a flag for 32-bit OpenCL on Windows would be bad either, though I don't think anybody will use it.

Hopefully, after that I will be able to get back to actually studying the language instead of build behavior.

**Comment by Athas** [comment]
I have attempted to make this change, but I'm not entirely sure what
System.Info.os looks like on Windows.  Please try it out!  I'll be gone
for a few hours.

-- 
\  Troels
/\ Henriksen

**Comment by mrakgr** [comment]
It compiles and runs correctly, but the timing is a bit off. fact 2000000000 runs in 15m my machine instead of 7ms like on Linux and sometimes the time actually gets logged as 0ms for some reason.

I haven't tested the 32-bit version yet, so I cannot tell whether this is because it is compiling 64-bit. I'll give it a try.

**Comment by wvdschel** [comment]
@mrakgr did you make a typo there? Is that supposed to be 15m(inutes) or 15ms?

**Comment by mrakgr** [comment]
It was a year since that post so my memory is not too great, but I think it was 15ms. I think the issue we discussed here were the floating point errors on the lower end in the timing printouts. On Windows it would switch between 15ms and 0ms and exactly those values because the values would get clipped somewhere. This is unlike on Linux where I would get precise measurements.

We fixed this, I think? Tell us if we did not.
